File: matlab/P2L2.m
function [Mxyev, Mxyiv, vb, Mzev, Mziv] = P2L2(vars,fdv)
%Format: function [Mxyev, Mxyiv, vb, Mev, Miv] = P2LBv4(vars,fdv)
%
%input:
    % vars:                 values for parms have been or will be fit
    % fdv = "forward variables" containing info about the measurement
    %    fdv.fitvars        Names of parms that need to be fit
    %    fdv.knowns         Names of parms that are known
    %    fdv.knownvals      Values of known parameters
    %    fdv.FlipAngles     Excitation angle list [2 x NFlips] in degrees
    %    fdv.TR:            Repetition time [2 x NFlips]
    %    fdv.taxis:         time axis for the data [1 x NFlips]
    %    fdv.data           [2 x NTP] observed [pyr;lac] data
    %    fdv.UseVIF         =1 if VIFs have been measured
    %    fdv.VIFP           **Mz** VIF for pyruvate [1 x NFlips]
    %    fdv.VIFL           **Mz** VIF for lactate [1 x NFlips] (usually zeros)
    %    fdv.verbose        =1 to plot results of this script
%    
%output:
    % Mxyev                 [2 x NTP] extravascular observed [Pyr;Lac] 
    % Mxyiv                 [2 x NTP] intravascular observed [Pyr;Lac]
    % vb                    vascular blood vol fraction
    % Mzev                   [2 x NTP] extravascular longitudinal [Pyr;Lac]
    % Mziv                   [2 x NTP] intravascular observed [Pyr;Lac]
%
% Note Mzev,Mziv reflects z-directred magnetization before the excitation pulse ,  
%  while Mxy reflects transverse magnetization due to pulse.
%
% Note total observed signal is [Pyr;Lac] = (1-vb)*Mxyev + vb*Mxyiv
%

    eps=1e-6;
       
    %Unpack variables:
    for ii=1:length(vars)
        x.(fdv.fitvars{ii})=vars(ii);
    end
    for ii=1:length(fdv.knowns)
        x.(fdv.knowns{ii})=fdv.knownvals(ii);
    end
    vb=x.vb;
    kvedve=x.kve/(1-vb);

    %Assume for now that the VIF is given:
    if fdv.UseVIF
        ff=[x.VIFScale*fdv.VIFP;
            x.VIFScale*fdv.VIFL];
    else 
        ff=[x.VIFScale*gampdf(fdv.taxis,x.Gam1,x.Gam2);
            zeros(1,fdv.NFlips)];
    end
    %IV magnetization at each segment:
    MivSeg =ff;
    MxyivSeg=ff.*sind(fdv.FlipAngle);
    
    %Initialize return variables:
    Mxyev=zeros(2,fdv.ntp);
    Mzev =zeros(2,fdv.ntp);
    Mxyiv=zeros(2,fdv.ntp);
    Mziv =zeros(2,fdv.ntp);

    %Manage trivial case of vb=0: no signal, or vb=1: all vessel
    if (vb<eps) 
        %All return variables have been initialized to zero, above.
        return
    elseif (vb>(1-eps))
        % If segmented acquisition with NFlips>NTP, where VIF has been
        % measured or interpolated to each excitation
        % Combine segments: ** NOTE HERE ASSUMES EQUAL WEIGHTING **
        for ii=1:fdv.ntp 
            Mxyiv(:,ii)=mean(MxyivSeg(:,(ii-1)*fdv.NSeg+1:ii*fdv.NSeg),2);
            Mziv(:,ii)=mean(MivSeg(:,(ii-1)*fdv.NSeg+1:ii*fdv.NSeg),2);
        end
        return
    end
    
    
    %Calculate nontrivial result
    % [Pev;Lev] = [Pev0;Lev0]*exp(At) +kpl*integral(0,t:exp(A*(t-tau))*[Piv(tau);Liv(tau)]*dtau)
    % A is 2x2
    a11=-((kvedve) + x.kpl + (1/x.T1Pyr));
    a12=x.klp;
    a21=x.kpl;
    a22=-((kvedve) + x.klp + (1/x.T1Lac));
    A = [a11 a12 ; a21 a22];
    % Diagonlize to permit matrix integral.  
    [P,D]=eig(A);
    % D is diagonlized matrix (2x2)
    % P is matrix of eigenvectors 
    %   --> A*P=P*D
    %   --> A=P*D\P
    dD=diag(D); % get diagonal values (1x2)
    
    %Calculate signal evolution each TR:
    
    MzevSegIC=[x.P0; x.L0]; %Set Initial Conditions
    for ii=1:fdv.ntp
        MxyevSeg=zeros(2,fdv.NSeg);
        MzevSeg =zeros(2,fdv.NSeg);
        for jj=1:fdv.NSeg
            iiseg=(ii-1)*(fdv.NSeg)+jj; % Works when NSeg=1 too
            TR=fdv.TR(iiseg);
            %
            %First account for EV signal already present and its evolution
            %
            %Longitudinal magnetization available at the start of each segment:
            MzevSeg(:,jj)=MzevSegIC;
            %Signal observed at each excitation, at start of segment:
            MxyevSeg(:,jj)=MzevSegIC.*sind(fdv.FlipAngle(:,iiseg));
            %At the end of this TR, Mz evolves to contribute to IC for next:
            if iiseg<fdv.NFlips % Don't need to calc after last datapoint
                MzevSeg1=(exp(dD*TR)).*(P\(MzevSegIC.*cosd(fdv.FlipAngle(:,iiseg))));
                %
                %Now calculate new spins flowing into the system
                %
                %Assume piecewise linear VIF. Diagonalize:
                dff1=P\ff(:,iiseg);    % Diag'd VIF @ start of TR
                dff2=P\ff(:,iiseg+1);  % Diag'd VIF @ end of TR
                %Get slope and y-intercept for diagonolized forcing function:
                b=dff1;
                m=(dff2-dff1)/TR;
                %At the end of this TR, inflowing spins will lead to:
                %MevSeg2a=exp(dD*TR).*(-b./dD).*(exp(-dD*TR)-1);
                %MevSeg2b=exp(dD*TR).*m.*(((-TR./dD)-(1./dD./dD)).*exp(-dD*TR)+(1./dD./dD));
                %slightly more efficient way to calculate:
                MevSeg2a=(-b./dD).*(1-exp(dD*TR));
                MevSeg2b=m.*(((-TR./dD)-(1./dD./dD))+exp(dD*TR).*(1./dD./dD));
                %Total signal at end of TR equals IC for next TR:
                MzevSegIC = P*(MzevSeg1 + kvedve*(MevSeg2a+MevSeg2b));
               
            end
        end
        if (fdv.NSeg>1) %manage signal from segmented acquisitions
            %sizemxyev=size(mean(MxyevSeg))
            Mxyev(:,ii)=mean(MxyevSeg,2);
            Mzev(:,ii) =mean(MzevSeg,2);
            Mxyiv(:,ii)=mean(MxyivSeg(:,(ii-1)*fdv.NSeg+1:ii*fdv.NSeg),2);
            Mziv(:,ii) =mean(MivSeg(:,(ii-1)*fdv.NSeg+1:ii*fdv.NSeg),2);
        else
            Mxyev(:,ii)=MxyevSeg;
            Mzev(:,ii) =MzevSeg;
            Mxyiv(:,ii)=MxyivSeg(:,ii);
            Mziv(:,ii) =MivSeg(:,ii);
        end            
    end

if fdv.verbose
    figure(99)
    plot(fdv.taxis(1:fdv.NSeg:end),Mxyev(1,:),'g-',...
        fdv.taxis(1:fdv.NSeg:end),Mxyev(2,:),'b-')
    legend('PyrEV','LacEV')
end    

File: python/P2L2.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats

import helperFunctions as hf

def P2L2(vars, fdv):
    '''
    input:
        vars:               Value for parameter to be fit as array of size (2,)
        fdv:                Dictionary with the following keys
            fdv['fitvarNames']      Names of variables to be fit as string
            fdv['knowns']           Dictionary object containing known values
                                        (see TestP2L1.py for default)
            fdv['ntp']              Number of time points as int
            fdv['NSeg']             Number of segments per time point as int
                                        Note: for version 0.0.1, NSeg = 1
            fdv['NFlips']           Total number of excitations = ntp * NSeg as int
            fdv['FlipAngle']        Numpy array of excitation angles in degrees size = (2, NFlips)
            fdv['TR']               Single valued numpy array of repetition time size = (1, NFlips)
            fdv['data']             Observed data in the format [ [pyr], [lac] ] with size = (2, ntp)
            fdv['taxis']            Time axis used for the data with size = (1, NFlips)
            fdv['UseVIF']           Bool type. True if VIFs have been measured
            fdv['VIFP']             **Mz** VIF for pyruvate with shape = (1, NFlips)
            fdv['VIFL']             **Mz** VIF for lactate with shape = (1, NFlips), usually zeros
            fdv['Name']             Title for the output. Only affects plot title on TestP2L1.py
            fdv['verbose']          Bool type. True plots results of P2L1 directly

    output:
        Mxyev:              Extravascular observed with shape = (2, NTP) as [ [pyr], [lac] ]
        Mxyiv:              Intravascular observed with shape = (2, NTP) as [ [pyr], [lac] ]    
        vb:                 vascular blood vol fraction
        Mzev:               Extravascular longitudinal with shape = (2, NTP) as [ [pyr], [lac] ]
        Mziv:               Intravascular longitudinal with shape = (2, NTP) as [ [pyr], [lac] ]

    Note: Mzev, Mziv reflects the z-directed magnetization before the excitation pulse,
        while Mxy reflects transverse magnetization due to pulse

    Total observed signal is [Pyr, Lac] (1-vb)*Mxyev + vb*Mxyiv
    '''

    eps = 1e-6

    # Unpack fit variables into better naming scheme
    if len(vars) != 2:
        print(
            'P2L2 requires two parameters to be fit. Please enter vars as list with shape (1, 2)'
        )
        return 1

    i = 0
    for item in fdv['fitvarNames']:
        fdv[item] = vars[i]
        i += 1

    kvedve = fdv['knowns']['kve'] / (1 - fdv['knowns']['vb'])

    # Assuming that the VIF is given
    if fdv['UseVIF']:
        ff = np.array( [ fdv['VIFScale']*fdv['VIFP'], fdv['VIFScale']*fdv['VIFL'] ] )
    else:
        ff = np.array( [ fdv['VIFScale']*stats.gamma(
            fdv['taxis'], fdv['knowns']['Gam1'], fdv['knowns']['Gam2']), np.zeros(1, fdv['FlipAngle']) ] )

    # IV magnetization at each segment
    MivSeg = ff
    MxyivSeg = ff *  np.sin( np.radians(fdv['FlipAngle']) )

    # Initialize return variables
    Mxyev = np.zeros( (2,fdv['ntp']) )
    Mzev = np.zeros( (2,fdv['ntp']) )
    vb = fdv['knowns']['vb']
    Mxyiv = np.zeros( (2,fdv['ntp']) )
    Mziv = np.zeros( (2,fdv['ntp']) )

    #Check for trivial cases
    if (vb < eps):
        print('\nNo signal acquired\n')
        return (Mxyev, Mxyiv, vb, Mzev, Mziv)
    elif vb > (1-eps):
        print('\nAll vessel signal\n')
        # If segmented scquisition with NFlips > NTP, where VIF has
        #   been measured or interpolated to each excitation
        #   combine segments: ** THIS ASSUMES EQUAL WEIGHTING **
        for i in range(0, fdv['ntp']):
            Mxyiv[:,i] = np.mean( MxyivSeg[:, (i-1)*fdv['Nseg']+1 : i*fdv['NSeg']], 2 )
            Mziv[:, i] = np.mean(MivSeg[:, (i-1)*fdv['NSeg']+1 : i*fdv['NSeg']], 2)
        return 

    # Calculate non trivial result:
    # [Pev, Lev] = [Pev0, Lev0]*exp(At)+kpl*integral(0,t:exp(A*(t-T))*Piv(T), Liv(T)*dT)
    # A is a matrix with shape (2,2)
    a11 = - ( kvedve + fdv['kpl'] + (1/fdv['knowns']['T1Pyr']) )
    a12 = fdv['knowns']['klp']
    a21 = fdv['kpl']
    a22 = - ( kvedve + fdv['knowns']['klp'] + (1/fdv['knowns']['T1Lac']) )
    A = [ [a11, a12], [a21, a22] ]
    
    dD, P = np.linalg.eig(A)
    # dD should be array of eigenvalues with shape (1,2)
    # P is a matrix containing the eigenvectors of A

    ### Calculate signal evolution in TR ###
    MzevSegIC = np.array( [fdv['knowns']['P0'], fdv['knowns']['L0']] ) # Set initial conditions
    for i in range(fdv['ntp']):
        MxyevSeg = np.zeros( (2, fdv['NSeg']) )
        MzevSeg = np.zeros( (2, fdv['NSeg']) )
        for j in range(fdv['NSeg']):
            iSeg = (i-1)*(fdv['NSeg']+j)
            TR = fdv['TR'][0][iSeg]

            # First account for signal already present and its evolution
            #  Longitudinal magnetization available at the start of each segment
            MzevSeg[:,j] = MzevSegIC
            # Signal observed at each excitation, at start of segment
            MxyevSeg[:,j] = np.multiply(
                MzevSegIC, np.sin( np.radians(fdv['FlipAngle'][:,iSeg]) )
            )

            # At the end of this TR, Mz evolves to contribute to IC for next
            if iSeg<fdv['NFlips']: # Don't calculate after last datapoint
                MzevSeg1 = np.exp(dD*TR).flatten() * hf.mrdivide(P, np.multiply(
                        MzevSegIC, np.cos( np.radians(fdv['FlipAngle'][:, iSeg]) )
                    ).T ).flatten()


            # Now calculate new spins flowing into the system
            #  Assume piecewise linear VIF and diagonalize
            dff1 = hf.mrdivide(P, ff[:,iSeg]) # Diag VIF @ start of TR
            dff2 = hf.mrdivide(P, ff[:,iSeg+1]) # Diag VIF @ end of TR            
            
            # Get slope and y-intercept for diagonalized pyruvate forcing function
            b = dff1
            m = (dff2-dff1) / TR

            # At end of this TR inflowing spins will cause
            MevSeg2a = - (b.flatten() / dD) * (1-np.exp(dD*TR))
            MevSeg2b = np.multiply(m.flatten(),
                (-np.divide(TR, dD)- ((1/dD)/dD) ) + (np.exp(dD*TR) * ((1/dD)/dD)) )
            
            # Total signal at end of TR is combination of inflowing and already present signals
            MzevSegIC = np.matmul(P, (MzevSeg1 + kvedve*(MevSeg2a+MevSeg2b)))



        if fdv['NSeg'] > 1:
            # In version 0.0.1, fdv['NSeg']>1 is not supported
            print('Script does not work for segmented acquisition yet')
        else:
            Mxyev[:,i-1] = np.reshape(MxyevSeg, (2,) )
            Mzev[:,i-1] = np.reshape(MzevSeg, (2,) )

            Mxyiv[:,i-1] = MxyivSeg[:,i-1]
            Mziv[:,i-1] = MivSeg[:,i-1]
    ### END OF CALCULATION LOOP ###

    if fdv['verbose']:
        fig, ax = plt.subplots(figsize = (12, 6))

        ax.plot(fdv['taxis'], Mxyev[0,:], color = 'green')
        ax.plot(fdv['taxis'], Mxyev[1,:], color = 'blue')

        ax.grid()

        plt.show()

    return (Mxyev, Mxyiv, vb, Mzev, Mziv)